ops/pebble.py

```diff
@@ -43,22 +43,243 @@
 import urllib.parse
 import urllib.request
 import warnings
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    BinaryIO,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    TextIO,
+    Tuple,
+    TypeVar,
+    Union,
+)
 
 from ops._private import yaml
 from ops._vendor import websocket
 
+if TYPE_CHECKING:
+    _Boundary = Tuple[int, int, bool]
+    from email.message import Message
+
+    from typing_extensions import Literal, Protocol, TypedDict
+
+    # callback types for _MultiParser header and body handlers
+    class _BodyHandler(Protocol):
+        def __call__(self, data: bytes, done: bool = False) -> None: ...  # noqa
+
+    _HeaderHandler = Callable[[bytes], None]
+    _StrOrByt = Union[str, bytes]
+    Numerical = Union[float, int]
+
+    # tempfile.NamedTemporaryFile has an odd interface because of that
+    # 'name' attribute, so we need to make a Protocol for it.
+    class _Tempfile(Protocol):
+        name = ''
+        def write(self, data: bytes): ...  # noqa
+        def close(self): ...  # noqa
+
+    class _FileLikeIO(Protocol[typing.AnyStr]):  # That also covers TextIO and BytesIO
+        def read(self, __n: int = ...) -> typing.AnyStr: ...  # for BinaryIO  # noqa
+        def write(self, __s: typing.AnyStr) -> int: ...  # noqa
+        def __enter__(self) -> typing.IO[typing.AnyStr]: ...  # noqa
+
+    class _Readable(Protocol):
+        def read(self, n: int = -1) -> _StrOrByt: ...  # noqa
+
+    class _Writable(Protocol):
+        # We'd need something like io.ReadableBuffer here,
+        # but we can't import that type
+        def write(self, buf: Union[bytes, str, bytearray]) -> int: ...  # noqa
+
+    _AnyStrFileLikeIO = typing.Union[_FileLikeIO[bytes], _FileLikeIO[str]]
+    _IOSource = Union[str, bytes, _AnyStrFileLikeIO]
+    _PartType = Literal['response', 'files']
+    _Ctype = Literal['multipart/form-data', 'application/json']
+    _Level = Optional[Union['CheckLevel', str]]
+    _Status = Union['CheckStatus', str]
+    _WarningDict = TypedDict('_WarningDict',
+                             {'message': str,
+                              'first-added': str,
+                              'last-added': str,
+                              'last-shown': Optional[str],
+                              'expire-after': str,
+                              'repeat-after': str})
```

Comment:
It might be that if we move the types to the global namespace (indeed get rid of the `if TYPE_CHECKING` nesting...) then they will get parsed and linked?
The type defs are valid python code, I am used to separating the types with those nested blocks for several reasons. I'll see if I can keep them closer to where they're used and if by doing so they show up in readthedocs.